### PR TITLE
Master bdg

### PR DESCRIPTION
**Pull request by micchie**
1. Patch series to to make vale more modular for upcoming paste integration.
2. Option to decide the number of host rings after config to enable an option that ensures host rings exactly parallel to hw rings.
3. Restore the private mem with proper reference counting on failure or deactivation of bwrap.

### CODE REVIEW

File: sys/dev/netmap/netmap.c [hunk 1]
```diff
@@ -977,17 +977,25 @@ netmap_hw_krings_delete(struct netmap_adapter *na)
...
+		netmap_mem_put(na->nm_mem);
+		na->nm_mem = na->nm_mem_prev;
+		na->nm_mem_prev = NULL;
+	}
+}
+
 static void
 netmap_mem_drop(struct netmap_adapter *na)
 {
 	int last = netmap_mem_deref(na->nm_mem, na);
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap.c:993]: This call should be removed, since you are doing it below.

File: sys/dev/netmap/netmap.c [hunk 2]
```diff
@@ -2917,19 +2925,13 @@ netmap_ioctl(struct netmap_priv_d *priv, u_long cmd, caddr_t data,
 			NMG_UNLOCK();
 			break;
 		}
-#ifdef WITH_VALE
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap.c:2920]: Why the removal of the ifdefs?

File: sys/dev/netmap/netmap_bdg.c [hunk 3]
```diff
@@ -1585,6 +1744,10 @@ netmap_bwrap_attach_common(struct netmap_adapter *na,
 		na->na_flags |= NAF_HOST_RINGS;
 		hostna = &bna->host.up;
 
+		/* bwrap_config() called in update_config() might
+		 * increase these default ring parameters.
+		 */
+
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap_bdg.c:1750]: Is there any problem in just moving all the ring parameters settings to ```bwrap_config```?

File: sys/dev/netmap/netmap_kern.h [hunk 4]
```diff
@@ -758,6 +758,7 @@ struct netmap_adapter {
 /* free */
 #define NAF_MOREFRAG	512	/* the adapter supports NS_MOREFRAG */
 #define NAF_OFFSETS	1024	/* the adapter supports the slot offsets */
+#define NAF_HOST_ALL	2048	/* the adapter wants as many host rings as hw */
```
Comment (giuseppelettieri) [inline sys/dev/netmap/netmap_kern.h:761]: What happens if the user asks for less/more?

**Comment by micchie** [comment]
I updated to apply your comments.

**Pull request merged**